### DISCUSSION TITLE
Strict validation of different date (and time) formats

### DISCUSSION THREAD

**Discussion by repli2dev** [CuyZ/Valinor#350] [Q&A] [answered]
In my APIs we use the full (`DATE_ATOM`) dates and occasionally just the `Y-m-d` dates...

The library provides `supportDateFormats` to make this work, though this does allow passing a full date time into the fields that should only have `Y-m-d` format...

From the docs it is possible to add custom constructor `registerConstructor` but  when I try to use it with `DateOnly` (descendant of `DateTimeImmutable` I will get `A collision was detected between the following constructors of the class`.

What is a supposed way of doing this if I don't want to disconnect the `DateTimeImmutable` and `DateOnly` classes?

**Comment by romm** [comment]
Hi @repli2dev, sorry for the late answer. I'll need more details, could you post what `DateOnly` looks like (especially its constructors signature), and how you mapper is built?

**Reply by repli2dev** [reply]
It was just a simple `class DateOnly extends DateTimeImmutable` and with mapper I tried:
1. `infer(DateTimeInterface::class, fn () => DateOnly::class)`
2. `registerConstructor(function (string $date): DateOnly {...}, DateOnly::class)`
3. Both

**Comment by repli2dev** [comment]
Sadly I lost the playground with the code, I tried to reconstruct:
```
class DateOnly extends DateTimeImmutable {
    public function __construct(string $time = 'now') {
        parent::__construct($time);
        $this->setTime(0, 0, 0, 0);
    }

    public function __toString(): string {
        return $this->format('Y-m-d');
    }

}
final readonly class MyClass
{
    public string $name;

    public DateOnly $date;

    public DateTimeInterface $date2;
}

$json = <<<JSON
    {
        "name": "France",
        "date": "2021-01-01T02:02:03+10:00",
        "date2": "2021-01-01T01:02:03+10:00"
    }
JSON;

try {
    $foo = (new \CuyZ\Valinor\MapperBuilder())
        ->supportDateFormats( DATE_ATOM, 'Y-m-d')
        ->mapper()
        ->map(MyClass::class, \CuyZ\Valinor\Mapper\Source\Source::json($json));

    var_dump($foo);
} catch (\CuyZ\Valinor\Mapper\MappingError $error) {
    throw $error;
}
```

This provides undesired time part in the `date` field
```
class MyClass#136 (3) {
  public readonly string $name =>
  string(6) "France"
  public readonly DateOnly $date =>
  class DateOnly#360 (3) {
    public $date =>
    string(26) "2021-01-01 02:02:03.000000"
    public $timezone_type =>
    int(1)
    public $timezone =>
    string(6) "+10:00"
  }
  public readonly DateTimeInterface $date2 =>
  class DateTimeImmutable#656 (3) {
    public $date =>
    string(26) "2021-01-01 01:02:03.000000"
    public $timezone_type =>
    int(1)
    public $timezone =>
    string(6) "+10:00"
  }
}
```

No change with 
```
->infer(DateOnly::class, fn () => DateOnly::class)
```

Not even with
```
->registerConstructor(function (string $time): DateOnly {
            return new DateOnly($time);
        })
```

What I need is to ensure that the DateOnly value is treated with its constructor and other `DateTimeInterface` through theirs.

**Comment by romm** [comment, marked as answer]
Hi @repli2dev, maybe I missed something, or maybe something was fixed in the library in the previous months, but the following code should give you what you want to achieve:

```php
class DateOnly extends DateTimeImmutable
{
    public function __toString(): string
    {
        return $this->format('Y-m-d');
    }
}

final readonly class MyClass
{
    public string $name;

    public DateOnly $date;

    public DateTimeInterface $date2;
}

$json = <<<JSON
    {
        "name": "France",
        "date": "2021-01-01T02:02:03+10:00",
        "date2": "2021-01-01T01:02:03+10:00"
    }
JSON;

try {
    $foo = (new \CuyZ\Valinor\MapperBuilder())
        ->supportDateFormats(DATE_ATOM, 'Y-m-d')
        ->registerConstructor(function (string $date): DateOnly {
            return (new DateOnly($date))->setTime(hour: 0, minute: 0, second: 0, microsecond: 0);
        })
        ->mapper()
        ->map(MyClass::class, \CuyZ\Valinor\Mapper\Source\Source::json($json));

    var_dump($foo);
} catch (\CuyZ\Valinor\Mapper\MappingError $error) {
    throw $error;
}
```

I'll be closing the issue; if it does not answer your needs please feel free to reopen!

**Reply by repli2dev** [reply]
😮‍💨 I got tricked by PHP again, thanks for the insight.
